keras/backend/mlx/core.py

```diff
@@ -0,0 +1,265 @@
+        def to_scalar_list(x):
+            if isinstance(x, list):
+                return [to_scalar_list(xi) for xi in x]
+            elif isinstance(x, mx.array):
+                if x.ndim == 0:
+                    return x.item()
+                else:
+                    return x.tolist()
+            else:
+                return x
+        return mx.array(to_scalar_list(x), dtype=mlx_dtype)
+
+    return mx.array(x, dtype=mlx_dtype)
+
+
+def convert_to_numpy(x):
+    # Performs a copy. If we want 0-copy we can pass copy=False
+    return np.array(x)
+
+
+def is_tensor(x):
+    return isinstance(x, mx.array)
+
+
+def shape(x):
+    return tuple(x.shape)
+
+
+def cast(x, dtype):
+    return convert_to_tensor(x, dtype=dtype)
+
+
+# Shape / dtype inference util
+def compute_output_spec(fn, *args, **kwargs):
+    def has_none_shape(x):
+        """Check for if a `KerasTensor` has dynamic shape."""
+        if isinstance(x, KerasTensor):
+            return None in x.shape
+        return False
+
+    def convert_keras_tensor_to_mlx(x, fill_value=None):
+        """Convert `KerasTensor`s to `mlx.array`s."""
+        if isinstance(x, KerasTensor):
+            shape = list(x.shape)
+            if fill_value:
+                for i, e in enumerate(shape):
+                    if e is None:
+                        shape[i] = fill_value
+            return mx.ones(shape, dtype=MLX_DTYPES[x.dtype])
+        return x
+
+    def convert_mlx_to_keras_tensor(x):
+        """Convert `mlx.array`s to `KerasTensor`s."""
+        if is_tensor(x):
+            return KerasTensor(x.shape, standardize_dtype(x.dtype))
+        return x
+
+    def symbolic_call(fn, args, kwargs, fill_value):
+        """Call `fn` to infer output shape and dtype."""
+        arr_args, arr_kwargs = tree.map_structure(
+            lambda x: convert_keras_tensor_to_mlx(x, fill_value),
+            (args, kwargs),
+        )
+        return fn(*arr_args, **arr_kwargs)
+
+    with StatelessScope():
+        outputs = symbolic_call(fn, args, kwargs, fill_value=83)
+
+        none_in_shape = any(map(has_none_shape, tree.flatten((args, kwargs))))
+        if none_in_shape:
+            outputs_1 = outputs
+            outputs_2 = symbolic_call(fn, args, kwargs, fill_value=89)
+
+            flat_out_1 = tree.flatten(outputs_1)
+            flat_out_2 = tree.flatten(outputs_2)
+
+            flat_out = []
+            for x1, x2 in zip(flat_out_1, flat_out_2):
+                shape = list(x1.shape)
+                for i, e in enumerate(x2.shape):
+                    if e != shape[i]:
+                        shape[i] = None
+                flat_out.append(KerasTensor(shape, standardize_dtype(x1.dtype)))
+            outputs = pack_sequence_as(outputs_1, flat_out)
+
+        output_spec = tree.map_structure(convert_mlx_to_keras_tensor, outputs)
+    return output_spec
+
+
+def cond(pred, true_fn, false_fn):
+    # TODO: How should we avoid evaluating pred in case we are tracing?
```

Comment:
Tracing may be the wrong word here.

Since mlx is lazily evaluated, in the trainer we can just run everything and if we don't `mx.eval()` then nothing will actually be computed.

Other cases that a graph is computed of course are printing or casting to a python value. In this case, since we need to compute the `bool` value for the condition we would evaluate the graph up until `pred` which may cause problems down the line.